Controllers/HierarchyTerritoriesAdminController.cs

```diff
@@ -0,0 +1,525 @@
+using Orchard.Core.Contents.Settings;
+using Orchard.Data;
+using Orchard.Environment.Extensions;
+using Orchard.Localization;
+using Orchard.Logging;
+using Orchard.Mvc.Extensions;
+using Orchard.Mvc.Html;
+using Orchard.Security;
+using Orchard.Security.Permissions;
+using Orchard.UI.Admin;
+using Orchard.UI.Notify;
+using Orchard.Utility.Extensions;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Web.Mvc;
+using System.Web.Routing;
+
+namespace Nwazet.Commerce.Controllers {
+    [OrchardFeature("Territories")]
+    [Admin]
+    [ValidateInput(false)]
+    public class HierarchyTerritoriesAdminController : Controller, IUpdateModel {
+
+        private readonly IContentManager _contentManager;
+        private readonly IContentDefinitionManager _contentDefinitionManager;
+        private readonly ITerritoriesService _territoriesService;
+        private readonly IAuthorizer _authorizer;
+        private readonly IWorkContextAccessor _workContextAccessor;
+        private readonly RouteCollection _routeCollection;
+        private readonly ITerritoriesHierarchyService _territoriesHierarchyService;
+        private readonly ITransactionManager _transactionManager;
+        private readonly INotifier _notifier;
+        private readonly IEnumerable<IContentHandler> _handlers;
+
+        public HierarchyTerritoriesAdminController(
+            IContentManager contentManager,
+            IContentDefinitionManager contentDefinitionManager,
+            ITerritoriesService territoriesService,
+            IAuthorizer authorizer,
+            IWorkContextAccessor workContextAccessor,
+            RouteCollection routeCollection,
+            ITerritoriesHierarchyService territoriesHierarchyService,
+            ITransactionManager transactionManager,
+            INotifier notifier,
+            IEnumerable<IContentHandler> handlers) {
+
+            _contentManager = contentManager;
+            _contentDefinitionManager = contentDefinitionManager;
+            _territoriesService = territoriesService;
+            _authorizer = authorizer;
+            _workContextAccessor = workContextAccessor;
+            _routeCollection = routeCollection;
+            _territoriesHierarchyService = territoriesHierarchyService;
+            _transactionManager = transactionManager;
+            _notifier = notifier;
+            _handlers = handlers;
+
+            T = NullLocalizer.Instance;
+            Logger = NullLogger.Instance;
+
+            _allowedTerritoryTypes = new Lazy<IEnumerable<ContentTypeDefinition>>(GetAllowedTerritoryTypes);
+            _allowedHierarchyTypes = new Lazy<IEnumerable<ContentTypeDefinition>>(GetAllowedHierarchyTypes);
+        }
+
+        public Localizer T;
+        public ILogger Logger { get; set; }
+
+        [HttpGet]
+        public ActionResult Index(int id) {
+            ActionResult redirectTo;
+            if (ShouldRedirectForPermissions(id, out redirectTo)) {
```

Comment:
Not on my dev machine it cannot. Maybe because I am using VS2015 (vanilla, without fiddling with it and telling it to use more modern compilers or stuff like that).